[WIP] Fix services deletion with finalizers

aojea opened the pull request:
> service IPs and ports are not released when deleting a service via a finalizer-removing update

Linking Jordan explanation of the Service deletion process in the apiserver

https://github.com/kubernetes/kubernetes/issues/87603#issuecomment-579443444

The problem is that meanwhile the service has not been deleted, the delete logic has already deleted the resources, but the repair loop recover them.

```
E0617 16:49:00.777663  500446 repair.go:240] the cluster IP [IPv4]:10.0.0.20 for service svc/default is not allocated; repairing
```

```
     services_test.go:127: unexpected error creating service: Service "svc" is invalid: spec.clusterIPs: Invalid value: 
 []string{"10.0.0.20"}: failed to allocated ip:10.0.0.20 with error:provided IP is already allocated
   services_test.go:129: Created service: svc
```


/kind bug
```release-note
fix a bug where services with finalizers weren't able to release the ClusterIP and/or NodePort resources allocated
```

Fixes: https://github.com/kubernetes/kubernetes/issues/87603

aojea committed:
integration test services finalizers

thockin committed:
Add stub begin* hooks in svc

thockin committed:
De-layer Service REST: Create

Gut the "outer" Create() and move it to the inner BeginCreate().  This
uses a "transaction" type to make cleanup functions easy to read.

thockin committed:
Service REST cleanup: Encapsulate allocator logic

Encapsulate the allocator logic so it can be shared across REST
layers while I stage a series of commits to get rid of one layer.

thockin committed:
Service REST cleanup: move methods -> alloc object

Move all allocator-related methods onto the alloc object so it can be
used in either REST layer.  There's an INORDINATE amount of test code
here and I am skeptical that it is all useful.  That's for later.

thockin committed:
Cleanup: Don't call validation directly from REST

The validation is called soon after anyway.

thockin committed:
FIXME: Comments for later cleanup

thockin committed:
Rename long, hard function name

thockin committed:
Fix up generic REST tests for service

thockin committed:
Overhaul Service REST Create test wrt dual-stack

This includes a few cases.

1) TestCreateIgnoresIPFamilyForExternalName: Prove that ExternalName is
ignored for dual-stack.  A small set of test cases were chosen to
demonstrate.

2) TestCreateIgnoresIPFamilyWithoutDualStack: Prove that when the
dual-stack gate is off, all services are ignored for dual-stack.  A
small set of test cases were chosen to demonstrate

3) TestCreateInitIPFields: Run over a huge array of test cases for
dual-stack.  This was generated by this program: https://gist.github.com/thockin/cccc9c9a580b4830ee0946ddd43eeafe

thockin committed:
Dedup svc tests for defaults

thockin committed:
Add dry-run support to the IP allocator subsystem

thockin committed:
Set IPs during dry-run service Create()

thockin committed:
Add test proving ExternalName does not set IPs

thockin committed:
Add test proving Headless does not set IPs

thockin committed:
De-layer Service REST: Delete

Gut the "outer" Delete() and move it to the inner AfterDelete().

thockin committed:
De-layer Service REST: Update

thockin committed:
Move svc ResourceLocation to 'inner' REST

Also move the test.

thockin committed:
Remove obviously unused args

Bridge the old tests into the new REST.

thockin committed:
WIP: only use 1 layer of REST in service

thockin committed:
Remove old, unused svc REST stubs

thockin committed:
Rename service NewGenericREST to NewREST

aojea committed:
move repair loop interval to a constant

aojea committed:
make repairinterable an option ONLY for tests

aojea commented on pkg/controlplane/controller.go:
```diff
@@ -110,11 +112,11 @@ func (c *completedConfig) NewBootstrapController(legacyRESTStorage corerest.Lega
 		SecondaryServiceClusterIPRange:    c.ExtraConfig.SecondaryServiceIPRange,
 
-		ServiceClusterIPInterval: 3 * time.Minute,
+		ServiceClusterIPInterval: c.RepairServicesInterval,
```
@liggitt is this allowed?
I need to expose this option to the test integration framework, otherwise the test has to last 3 mins

liggitt replied:
pretty sure we already have something similar for endpoint reconciler configuration (in ExtraConfig, not bound to a flag that I can see)

fedebongio commented:
/sig network
/remove-sig api-machinery

liggitt commented on pkg/controlplane/instance.go:
```diff
@@ -217,11 +219,15 @@ type ExtraConfig struct {
 type completedConfig struct {
 	GenericConfig genericapiserver.CompletedConfig
 	ExtraConfig   *ExtraConfig
+	// ONLY FOR TESTING Not exposed to the users
+	RepairServicesInterval time.Duration
```
this is fine to include, but put it in ExtraConfig

liggitt commented on pkg/controlplane/instance.go:
```diff
@@ -292,9 +298,13 @@ func (c *Config) createEndpointReconciler() reconcilers.EndpointReconciler {
+	}
 	cfg := completedConfig{
 		c.GenericConfig.Complete(c.ExtraConfig.VersionedInformers),
 		&c.ExtraConfig,
```
if this is in ExtraConfig, no need to change the signature here

aojea committed:
use Extraconfig to configure the repair interval

cbandy commented:
I'm interested in this PR because of my closed/duplicate issue #99407. There, I did not have to wait three minutes to see incorrect behavior regarding Endpoints.

fedebongio commented:
/remove-sig api-machinery

khenidak commented:
Based on this: https://github.com/kubernetes/kubernetes/issues/87603#issuecomment-579443444 we still don't have AfterDelete hook implementation for service object, right?

aojea commented:
> Based on this: [#87603 (comment)](https://github.com/kubernetes/kubernetes/issues/87603#issuecomment-579443444) we still don't have AfterDelete hook implementation for service object, right?



> Based on this: [#87603 (comment)](https://github.com/kubernetes/kubernetes/issues/87603#issuecomment-579443444) we still don't have AfterDelete hook implementation for service object, right?

@thockin does have it https://github.com/kubernetes/kubernetes/pull/96684/files :)

thockin committed:
Add a basic transaction API inside service REST

This will be used in upcoming commits.

thockin committed:
Move the allocations in service Create into funcs

All the logic remains unchanged, just reorganized.  The functions are
imperfect but emphasize the change being made and can be cleaned up
subsequently.

This makes the following steps easier to comprehend.

thockin committed:
Service REST test: better checks

"Has()" was strengthened in the older rest_test, now in the newer.

thockin committed:
Service REST test: remove redundant Create tests

These cases are all covered in storage_test.

thockin committed:
Service REST test: better NodePort tests

This commit ports the NodePort test from rest_test to storage_test.
It's not a direct port, though.  I have added many more cases (much more
exhaustive) and more assertions.

This includes cases for gate MixedProtocolLBService.

thockin committed:
WIP: Service: Use DeepCopy() and fix tests

If we don't DeepCopy() on the way into Create, we can end up writing
into the original object, which we then use to compare to the returned
object.  This works fine when we serialize across a network link but
hides errors when it's just memory.

WIP: this depends on #103545 which is not agreed upon yet.  If that does
not go in, then we need to DeepCopy() on calls to Create()

xref https://groups.google.com/g/kubernetes-sig-api-machinery/c/DZKBRdKIhmk/m/djddw5DgCAAJ

thockin committed:
REST: Use DeepCopy() on Create() and fix tests

If we don't DeepCopy() on the way into Create, we can end up writing
into the original object, which we then use to compare to the returned
object or fixed values.  This works fine when we serialize across a
network link but hides errors when it's just a shared object in memory.

thockin committed:
Service REST: Use DeepCopy() on Create() and fix tests

thockin committed:
Rename some tests for clarity

thockin committed:
Move test to reduce diffs in next commits

thockin committed:
Service REST test: HealthCheckNodePort tests

This commit ports the ExternalTrafficPolicy and HealthCheckNodePort
tests from rest_test to storage_test.  It's not a direct port, though.
I have added more cases (much more exhaustive) and more assertions.

thockin committed:
Remove unused NewContiguousAllocationMap

This was used at some point in the past and never removed.  We are not
in the business of hosting unused code.

thockin committed:
Allocator renames for clarity

Rename `NewCIDRRange()` to `NewInMemory()`
Rename `NewAllocatorCIDRRange()` to `New()`

Rename `NewPortAllocator()` to `NewInMemory()`
Rename `NewPortAllocatorCustom()` to `New()`

thockin committed:
Service REST test: IP and port reallocation

Make sure the logic that was covered in rest_test is covered in
storage_test.

thockin committed:
Service REST test: Remove redundant Get test

thockin committed:
Service REST test: clean up redundant delete tests

thockin committed:
Service REST: feature tests

This scaffolding allows us to assert more on each test case, and more
consistently.

Set input fields from output fields IFF they are expected AND not set on
input.  This allows us to verify the "after" state (expected) whether
the test case specified the value or not, and still pass the generic
cmp.Equal.

Some of the existing tests that are focused on create and delete can
probably be replaced by these, maybe.

This could be used in other test cases that are open-coding a lot of the
same stuff.  Later.

thockin committed:
Move ETP clearing to dropTypeDependentFields

thockin committed:
Service REST: add a delete-with-finalizer test

thockin committed:
ServiceREST: Use "prove" helpers in other tests

thockin committed:
Fix validation on ETP: "" is not valid

This was causing tests to pass which ought not be passing.

thockin committed:
Replace ETP create test with a feature test

All the same test cases and more.

thockin committed:
Add a test for PatchAllocatedValues

thockin committed:
Set IPs during dry-run service Update()

Also add a test.

aojea commented:
cool, this indeed solves the bug of services and finalizers
```
k8s.io/kubernetes/test/integration/network: TestServicesFinalizers
```

fedebongio commented:
/remove-sig api-machinery

thockin commented:
I'm closing this in favor of https://github.com/kubernetes/kubernetes/pull/96684 which also solves it.
